devdoc/channel.md

```diff
@@ -0,0 +1,297 @@
+
+**SRS_CHANNEL_43_081: [** `channel_create` shall call `sm_create`. **]**
+
+**SRS_CHANNEL_43_082: [** `channel_create` shall call `sm_open_begin`. **]**
+
+**SRS_CHANNEL_43_083: [** `channel_create` shall call `sm_open_end`. **]**
+
+**SRS_CHANNEL_43_084: [** `channel_create` shall call `DList_InitializeListHead`. **]**
+
+**SRS_CHANNEL_43_085: [** `channel_create` shall call `interlocked_exchange` to set the `state` to `CHANNEL_EMPTY`. **]**
+
+**SRS_CHANNEL_43_086: [** `channel_create` shall succeed and return the created `THANDLE(CHANNEL)`. **]**
+
+**SRS_CHANNEL_43_002: [** If there are any failures, `channel_create` shall fail and return `NULL`. **]**
+
+
+### channel_dispose
+```c
+    static void channel_dispose(CHANNEL* channel);
+```
+
+`channel_dispose` disposes the given `channel`.
+
+**SRS_CHANNEL_43_087: [** `channel_dispose` disposes the given `channel`. **]**
+
+**SRS_CHANNEL_43_088: [** `channel_dispose` shall `sm_close_begin_with_cb` with `channel_close` as the callback. **]**
+
+**SRS_CHANNEL_43_089: [** `channel_dispose` shall `sm_close_end`. **]**
+
+**SRS_CHANNEL_43_090: [** `channel_dispose` shall `sm_destroy`. **]**
+
+**SRS_CHANNEL_43_099: [** `channel_dispose` shall call `srw_lock_destroy`. **]**
+
+**SRS_CHANNEL_43_091: [** `channel_dispose` shall release the reference to `THANDLE(THREADPOOL)`. **]**
+
+**SRS_CHANNEL_43_092: [** `channel_dispose` shall release the reference to `THANDLE(CHANNEL_INTERNAL)`. **]**
+
+
+### channel_close
+```c
+    static void channel_close(void* context);
+```
+
+`channel_close` is the callback passed to `sm_close_begin_with_cb` by `channel_dispose`. It abandons all pending operations
+
+**SRS_CHANNEL_43_094: [** `channel_close` shall call `srw_lock_acquire_exclusive`. **]**
+
+**SRS_CHANNEL_43_095: [** `channel_close` shall iterate over the list of pending operations and do the following: **]**
+
+ - **SRS_CHANNEL_43_096: [** set the `operation state` to `CHANNEL_OP_ABANDONED`. **]**
+ 
+ - **SRS_CHANNEL_43_097: [** call `threadpool_schedule_work` with `execute_callbacks` as `work_function`. **]**
+
+**SRS_CHANNEL_43_100: [** `channel_close` shall call `srw_lock_release_exclusive`. **]**
+
+
+### channel_pull
+```c
+    MOCKABLE_FUNCTION(, CHANNEL_RESULT, channel_pull, THANDLE(CHANNEL), channel, PULL_CALLBACK, pull_callback, void*, pull_context, THANDLE(ASYNC_OP)*, out_op_pull);
+```
+
+`channel_pull` registers the given `pull_callback` to be called when there is data to be consumed.
+
+**SRS_CHANNEL_43_007: [** If `channel` is `NULL`, `channel_pull` shall fail and return `CHANNEL_RESULT_INVALID_ARGS`. **]**
+
+**SRS_CHANNEL_43_008: [** If `pull_callback` is `NULL`, `channel_pull` shall fail and return `CHANNEL_RESULT_INVALID_ARGS`. **]**
+
+**SRS_CHANNEL_43_009: [** If `out_op_pull` is `NULL`, `channel_pull` shall fail and return `CHANNEL_RESULT_INVALID_ARGS`. **]**
+
+**SRS_CHANNEL_43_010: [** `channel_pull` shall call `srw_lock_acquire_exclusive`. **]**
```

Comment:
so if the first thing that happens here is "acquire_exclusive" I don't understand why SM is then needed.

Consider removing SM and maybe using a simple counter variable?

Reply:
Removed SM